src/ring_theory/subsemiring/basic.lean

```diff
@@ -879,6 +879,10 @@ def cod_restrict (f : R →+* S) (s : σS) (h : ∀ x, f x ∈ s) : R →+* s :=
   .. (f : R →* S).cod_restrict s h,
   .. (f : R →+ S).cod_restrict s h }
 
+/--The ring homomorphism from the pullback of `s` to `s`. -/
+def restrict₂ (f : R →+* S) (s : subsemiring S) (h : ∀ (x : (subsemiring.comap f s)), f x ∈ s):
```

Comment:
This should be named `restrict`, and the old `restrict` should be named `dom_restrict`.
And I think it would be better to use `h : subsemiring.comap f s ≤ s` instead.

Reply:
OK thank you, I 've changed the names. 

However, both hypotheses  `(h : ∀ (x : (subsemiring.comap f s)), f x ∈ s)` and `h : subsemiring.comap f s ≤ s`  seem to be false for the map that I need, and the second one doesn't type check. 

What I need is the following:  If $f : X \rightarrow Y$ is a ring hom and $A$ is a subset of $Y$ then its pullback  `(comap A f)` will be $f^{−1}(A) = \{x\in X \mid f(x) \in A\}$, which is a subset of $X.$ 

Then I defined `ring_hom.valuation_subring_comap` which is  `(comap A f) →+* A`. 

I pushed the code, could you please take a look?

Reply:
Sorry. It should be
```
def restrict (f : R →+* S) (s' : subsemiring R) (s : subsemiring S) (h : s' ≤ s.comap f) :
  s' →+* s := (f.dom_restrict s').cod_restrict s (λ x, h x.2)
```